dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,221 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import traceback
+from multiprocessing import reduction
+from nvidia.dali.shared_batch import SharedMemChunk, SharedBatchMeta, SharedBatchWriter
+from nvidia.dali.messages import CompletedTasks
+
+
+class _ProcessedTasks:
+    """Internal worker message send to disptacher with completed tasks where it is
+    serialized and dispatched to the pool"""
+
+    def __init__(self, scheduled, mem_chunk=None, data_batch=None, exception=None,
+                 traceback_str=None):
+        self.context_i = scheduled.context_i
+        self.batch_i = scheduled.batch_i
+        self.mem_chunk = mem_chunk
+        self.data_batch = data_batch
+        self.exception = exception
+        self.traceback_str = traceback_str
+
+    @classmethod
+    def done(cls, scheduled, mem_chunk, data_batch):
+        return cls(scheduled, mem_chunk, data_batch)
+
+    @classmethod
+    def failed(cls, scheduled, exception, traceback_str=None):
+        return cls(scheduled, exception=exception, traceback_str=traceback_str)
+
+    def is_failed(self):
+        return self.exception is not None
+
+
+class SharedBatchesDispatcher:
+    """SharedBatchesDispatcher serializes batches, puts them into provided
+    shared memory chunks and notifies parent process of batch ready to be read from shared memory.
+    It keeps track of what shared memory chunks have been already sent and if needed, sends
+    file descriptors of memory chunks that parent process hasn't seen yet.
+
+    Parameters
+    ----------
+    `worker_id` : int
+        Id of the worker passed by the parent process. Added to messages sent over the ``res_pipe``
+        to simplify bookeeping in parent process.
+    `sock` : socket
+        Python wrapper around Unix socket, capable of sending file descriptors between processes.
+    `res_pipe`: pipe
+        Pipe used to send parent process a notification (along with essential meta data info) about
+        ready batch in a given shared memory chunk.
+    """
+
+    def __init__(self, worker_id, sock, res_pipe):
+        self.worker_id = worker_id
+        self.fd_sent = set()
+        self.sock = sock
+        self.res_pipe = res_pipe
+
+    def send(self, processed_tasks: _ProcessedTasks):
+        if processed_tasks.is_failed():  # one of the tasks failed
+            completed_tasks = CompletedTasks.failed(self.worker_id, processed_tasks)
+            self.res_pipe.send(completed_tasks)
+            return
+        mem_chunk = processed_tasks.mem_chunk
+        writer = SharedBatchWriter(processed_tasks.mem_chunk)
```

Comment:
```suggestion
        writer = SharedBatchWriter(mem_chunk)
```

Reply:
Maybe it can just accept mem_chunk nad data_batch in the constructor. Why to have this two separate steps which need to operate on the same processed_tasks anyway.

Reply:
Good idea. I think I will turn it into a function.